source/m2mtlvserializer.cpp

```diff
@@ -152,87 +155,64 @@ void M2MTLVSerializer::serialize_resource_instance(uint16_t id, M2MResourceInsta
 void M2MTLVSerializer::serialize_TILV(uint8_t type, uint16_t id, uint8_t *value, uint32_t value_length, uint8_t *&data, uint32_t &size)
 {
     uint8_t *tlv = 0;
-    uint32_t type_length = 1;
+    const uint32_t type_length = TLV_TYPE_SIZE;
     type += id < 256 ? 0 : ID16;
     type += value_length < 8 ? value_length :
             value_length < 256 ? LENGTH8 :
             value_length < 65536 ? LENGTH16 : LENGTH24;
-    uint8_t *tlv_type = (uint8_t*)malloc(type_length+1);
-    memset(tlv_type,0,type_length+1);
-    *tlv_type = type & 0xFF;
+    uint8_t tlv_type;
+    tlv_type = type & 0xFF;
 
-    uint32_t id_size = 0;
-    uint8_t* id_ptr = serialize_id(id, id_size);
+    uint32_t id_size;
+    uint8_t id_array[MAX_TLV_ID_SIZE];
+    serialize_id(id, id_size, id_array);
 
-    uint32_t length_size = 0;
-    uint8_t* length_ptr = serialize_length(value_length, length_size);
+    uint32_t length_size;
+    uint8_t length_array[MAX_TLV_LENGTH_SIZE];
+    serialize_length(value_length, length_size, length_array);
 
-    tlv = (uint8_t*)malloc(size + type_length + id_size + length_size + value_length+1);
-    memset(tlv,0,size + type_length + id_size + length_size + value_length+1);
+    tlv = (uint8_t*)malloc(size + type_length + id_size + length_size + value_length);
     if(data) {
         memcpy(tlv, data, size);
+        free(data);
     }
```

Comment:
I see that higher level serialize method already returns success information. I can make change, that malloc fail will be propagated as failure from serialize method. However that method is private, and public callers of the method don't currently handle success information. So in my understanding the change will prevent segfault, but causes silent failure and possible memory corruption later. Should we now return NULL from e.g. uint8_t* M2MTLVSerializer::serialize(M2MResource *resource, uint32_t &size) in "out of memory" situation?